html/plugins/Test/Test.php

```diff
@@ -2,26 +2,38 @@
 
 namespace LibreNMS\Plugins;
 
+use App\Models\Device;
+use App\Models\Port;
+
 class Test
 {
     public static function menu()
     {
-        echo '<li><a href="plugin/p=Test">Test</a></li>';
+        $name = self::className();
+        echo view(self::viewPath(), compact('name'));
     }
 
-    //end menu()
-
     public function device_overview_container($device)
     {
-        echo '<div class="container-fluid"><div class="row"> <div class="col-md-12"> <div class="panel panel-default panel-condensed"> <div class="panel-heading"><strong>' . get_class() . ' Plugin </strong> </div>';
-        echo '  Example plugin in "Device - Overview" tab <br>';
-        echo '</div></div></div></div>';
+        $name = self::className() . ' Plugin';
+        $device = Device::find($device['device_id']);
+        echo view(self::viewPath(), compact('name', 'device'));
     }
 
     public function port_container($device, $port)
     {
-        echo '<div class="container-fluid"><div class="row"> <div class="col-md-12"> <div class="panel panel-default panel-condensed"> <div class="panel-heading"><strong>' . get_class() . ' plugin in "Port" tab</strong> </div>';
-        echo 'Example display in Port tab</br>';
-        echo '</div></div></div></div>';
+        $name = self::className() . ' Plugin';
+        $port = Port::find($port['port_id']);
+        echo view(self::viewPath(), compact('name', 'port'));
+    }
+
+    private static function viewPath()
+    {
+        return 'plugins.' . strtolower(self::className()) . '.' . debug_backtrace()[1]['function'];
```

Comment:
It seems that each plugin can add its own view path, that could be done in a class constructor for instance.  
```
$path = base_path("html/plugins" . self::classname . "/ressources/views");
 \View::addLocation($path");
```

Goal is to keep the fact that a plugin is self_contained.

Of course, this view loading code should be in ~/LibreNMS/Plugins.php class, and expose the functions to the plugins themselves. That would ensure that no view collision occurs.

Reply:
@PipoCanaja thanks for help. I used a abstract class for Plugin.php can you have a look that you think about it?

Reply:
Will do my best to take the required amount of time ASAP to review this. A 10 seconds look at the files seems promising ! :)

Reply:
No rush! Whenever you find the time. For me it's never that urgent :-)